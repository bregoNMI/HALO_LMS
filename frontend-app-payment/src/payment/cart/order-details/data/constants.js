// Product types
export const COURSE_ENTITLEMENT_PRODUCT_TYPE = 'Course Entitlement';
export const ENROLLMENT_CODE_PRODUCT_TYPE = 'Enrollment Code';
export const SEAT_PRODUCT_TYPE = 'Seat';

// Certificate types
export const VERIFIED_CERTIFICATE_TYPE = 'verified';
export const CREDIT_CERTIFICATE_TYPE = 'credit';

// Order Details message keys
export const ENTITLEMENT_MESSAGE_KEY = 'entitlement';
export const ENROLLMENT_CODE_MESSAGE_KEY = 'enrollment.code';
export const VERIFIED_SEAT_MESSAGE_KEY = 'seat.verified';
export const CREDIT_SEAT_MESSAGE_KEY = 'seat.credit';
export const SEAT_MESSAGE_KEY = 'seat';
