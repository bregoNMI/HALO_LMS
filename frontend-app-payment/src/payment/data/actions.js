import { createRoutine } from 'redux-saga-routines';

// Routines are action + action creator pairs in a series.
// Actions adhere to the flux standard action format.
// Routines by default are in the form of:
//
// Action                |   Action Creator
// -----------------------------------------------
// fetchBasket.TRIGGER   |   fetchBasket()
// fetchBasket.SUCCESS   |   fetchBasket.success()
// fetchBasket.FAILURE   |   fetchBasket.failure()
// fetchBasket.FULFILL   |   fetchBasket.fulfill()
//
// Created with redux-saga-routines
export const fetchCaptureKey = createRoutine('FETCH_CAPTURE_KEY');
export const fetchClientSecret = createRoutine('FETCH_CLIENT_SECRET');
export const submitPayment = createRoutine('SUBMIT_PAYMENT');
export const fetchBasket = createRoutine('FETCH_BASKET');
export const addCoupon = createRoutine('ADD_COUPON');
export const removeCoupon = createRoutine('REMOVE_COUPON');
export const updateQuantity = createRoutine('UPDATE_QUANTITY');

// Actions and their action creators
export const BASKET_DATA_RECEIVED = 'BASKET_DATA_RECEIVED';

export const basketDataReceived = basket => ({
  type: BASKET_DATA_RECEIVED,
  payload: basket,
});

export const BASKET_PROCESSING = 'BASKET_PROCESSING';

export const basketProcessing = isProcessing => ({
  type: BASKET_PROCESSING,
  payload: isProcessing,
});

export const CAPTURE_KEY_PROCESSING = 'CAPTURE_KEY_PROCESSING';

export const captureKeyProcessing = isProcessing => ({
  type: CAPTURE_KEY_PROCESSING,
  payload: isProcessing,
});

export const CLIENT_SECRET_PROCESSING = 'CLIENT_SECRET_PROCESSING';

export const clientSecretProcessing = isProcessing => ({
  type: CLIENT_SECRET_PROCESSING,
  payload: isProcessing,
});

export const MICROFORM_STATUS = 'MICROFORM_STATUS';

export const microformStatus = status => ({
  type: MICROFORM_STATUS,
  payload: status,
});

export const CAPTURE_KEY_START_TIMEOUT = 'CAPTURE_KEY_START_TIMEOUT';

export const captureKeyStartTimeout = () => ({
  type: CAPTURE_KEY_START_TIMEOUT,
});
export const CAPTURE_KEY_DATA_RECEIVED = 'CAPTURE_KEY_DATA_RECEIVED';

export const captureKeyDataReceived = captureKey => ({
  type: CAPTURE_KEY_DATA_RECEIVED,
  payload: captureKey,
});

export const CLIENT_SECRET_DATA_RECEIVED = 'CLIENT_SECRET_DATA_RECEIVED';

export const clientSecretDataReceived = clientSecret => ({
  type: CLIENT_SECRET_DATA_RECEIVED,
  payload: clientSecret,
});
