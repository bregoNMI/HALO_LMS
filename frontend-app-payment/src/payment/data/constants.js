export const ORDER_TYPES = {
  BULK_ENROLLMENT: 'Enrollment Code',
  ENTITLEMENT: 'Course Entitlement',
  SEAT: 'Seat',
};

export const CERTIFICATE_TYPES = {
  VERIFIED: 'verified',
  CREDIT: 'credit',
};
