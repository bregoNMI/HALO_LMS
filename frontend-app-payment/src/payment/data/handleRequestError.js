import { logError, logInfo } from '@edx/frontend-platform/logging';
import { camelCaseObject } from './utils';

function handleFieldErrors(errors) {
  const fieldErrors = Object.entries(errors).map(([name, value]) => ({
    code: value.error_code ? value.error_code : null,
    userMessage: value.user_message,
    fieldName: name,
  }));

  const validationError = new Error();
  validationError.fieldErrors = fieldErrors;
  throw validationError;
}

function handleApiErrors(errors) {
  const apiErrors = errors.map(err => ({
    code: err.error_code ? err.error_code : null,
    userMessage: err.user_message ? err.user_message : null,
    messageType: err.message_type ? err.message_type : null,
  }));

  const apiError = new Error();
  apiError.errors = apiErrors;
  throw apiError;
}

function handleApiMessages(messages) {
  const apiError = new Error();
  apiError.messages = camelCaseObject(messages);
  throw apiError;
}

/**
 * Processes and re-throws request errors.  If the response contains a field_errors field, will
 * massage the data into a form expected by the client.
 *
 * If the response contains a single API error, will similarly format that for the client.
 *
 * Field errors will be packaged with a fieldErrors field usable by the client.
 *
 * @param error The original error object.
 * @param unpackFunction (Optional) A function to use to unpack the field errors as a replacement
 * for the default.
 */
export default function handleRequestError(error) {
  // Validation errors
  if (error.response && error.response.data.field_errors) {
    logInfo('Field Errors', error.response.data.field_errors);
    handleFieldErrors(error.response.data.field_errors);
  }

  // API errors
  if (error.response && error.response.data.errors !== undefined) {
    logInfo('API Errors', error.response.data.errors);
    handleApiErrors(error.response.data.errors);
  }

  // API messages
  if (error.response && error.response.data.messages !== undefined) {
    logInfo('API Messages', error.response.data.messages);
    handleApiMessages(error.response.data.messages);
  }

  // Single API error
  if (error.response && error.response.data.error_code) {
    logInfo('API Error', error.response.data.error_code);
    handleApiErrors([
      {
        error_code: error.response.data.error_code,
        user_message: error.response.data.user_message,
      },
    ]);
  }

  // SKU mismatch error
  if (error.response && error.response.data.sku_error) {
    logInfo('SKU Error', error.response.data.sku_error);
    handleApiErrors([
      {
        error_code: 'basket-changed-error-message',
        user_message: 'error',
      },
    ]);
  }

  // Basket already purchased
  if (error.code === 'payment_intent_unexpected_state' && error.type === 'invalid_request_error') {
    logInfo('Basket Changed Error', error.code);
    handleApiErrors([
      {
        error_code: 'basket-changed-error-message',
        user_message: 'error',
      },
    ]);
  }

  // Other errors
  logError(error);
  throw error;
}

// Processes API errors and converts them to error objects the sagas can use.
export function handleApiError(requestError) {
  try {
    // Always throws an error:
    handleRequestError(requestError);
  } catch (errorWithMessages) {
    const processedError = new Error();
    processedError.messages = errorWithMessages.messages;
    processedError.errors = errorWithMessages.errors;
    processedError.fieldErrors = errorWithMessages.fieldErrors;

    throw processedError;
  }
}
