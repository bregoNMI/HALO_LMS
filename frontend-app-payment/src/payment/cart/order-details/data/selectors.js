import { createSelector, createStructuredSelector } from 'reselect';

import { getModuleState } from '../../../data/utils';
import {
  SEAT_MESSAGE_KEY,
  CREDIT_SEAT_MESSAGE_KEY,
  VERIFIED_SEAT_MESSAGE_KEY,
  SEAT_PRODUCT_TYPE,
  VERIFIED_CERTIFICATE_TYPE,
  CREDIT_CERTIFICATE_TYPE,
  ENROLLMENT_CODE_PRODUCT_TYPE,
  COURSE_ENTITLEMENT_PRODUCT_TYPE,
  ENTITLEMENT_MESSAGE_KEY,
  ENROLLMENT_CODE_MESSAGE_KEY,
} from './constants';

export const storePath = ['payment', 'basket'];

const paymentSelector = (state, props) => getModuleState(state, props.storePath ? props.storePath : storePath);

const lastProductSelector = createSelector(
  paymentSelector,
  ({ products }) => products[products.length - 1],
);

const messageTypeSelector = createSelector(
  lastProductSelector,
  (lastProduct) => {
    if (lastProduct === undefined) {
      return null;
    }
    const { productType, certificateType } = lastProduct;
    if (productType === COURSE_ENTITLEMENT_PRODUCT_TYPE) {
      return ENTITLEMENT_MESSAGE_KEY;
    }
    if (productType === ENROLLMENT_CODE_PRODUCT_TYPE) {
      return ENROLLMENT_CODE_MESSAGE_KEY;
    }
    if (productType === SEAT_PRODUCT_TYPE) {
      if (certificateType === VERIFIED_CERTIFICATE_TYPE) {
        return VERIFIED_SEAT_MESSAGE_KEY;
      }
      if (certificateType === CREDIT_CERTIFICATE_TYPE) {
        return CREDIT_SEAT_MESSAGE_KEY;
      }
      return SEAT_MESSAGE_KEY;
    }
    return null;
  },
);

export const orderDetailsMapStateToProps = createStructuredSelector({
  messageType: messageTypeSelector,
});
