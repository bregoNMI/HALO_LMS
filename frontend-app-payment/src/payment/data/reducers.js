import { combineReducers } from 'redux';

import {
  BASKET_DATA_RECEIVED,
  BASKET_PROCESSING,
  CAPTURE_KEY_DATA_RECEIVED,
  CAPTURE_KEY_PROCESSING,
  CLIENT_SECRET_DATA_RECEIVED,
  CLIENT_SECRET_PROCESSING,
  MICROFORM_STATUS,
  fetchBasket,
  submitPayment,
  fetchCaptureKey,
  fetchClientSecret,
} from './actions';

import { DEFAULT_STATUS } from '../checkout/payment-form/flex-microform/constants';

const basketInitialState = {
  loading: true,
  loaded: false,
  submitting: false,
  redirect: false,
  isBasketProcessing: false,
  products: [],
};

const basket = (state = basketInitialState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case fetchBasket.TRIGGER: return { ...state, loading: true };
      case fetchBasket.FULFILL: return {
        ...state,
        loading: false,
        loaded: true,
      };

      case BASKET_DATA_RECEIVED: return { ...state, ...action.payload };

      case BASKET_PROCESSING: return {
        ...state,
        isBasketProcessing: action.payload,
      };

      case submitPayment.TRIGGER: return {
        ...state,
        paymentMethod: action.payload.method,
      };
      case submitPayment.REQUEST: return {
        ...state,
        submitting: true,
      };
      case submitPayment.SUCCESS: return {
        ...state,
        redirect: true,
      };
      case submitPayment.FULFILL: return {
        ...state,
        submitting: false,
        paymentMethod: undefined,
      };

      default:
    }
  }
  return state;
};

const captureContextInitialState = {
  isCaptureKeyProcessing: false,
  microformStatus: DEFAULT_STATUS,
  captureKeyId: '',
};

const captureKey = (state = captureContextInitialState, action = null) => {
  if (action !== null) {
    switch (action.type) {
      case fetchCaptureKey.TRIGGER: return state;
      case fetchCaptureKey.FULFILL: return state;

      case CAPTURE_KEY_DATA_RECEIVED: return { ...state, ...action.payload };

      case CAPTURE_KEY_PROCESSING: return {
        ...state,
        isCaptureKeyProcessing: action.payload,
      };

      case MICROFORM_STATUS: return {
        ...state,
        microformStatus: action.payload,
      };

      default:
    }
  }
  return state;
};

const clientSecretInitialState = {
  isClientSecretProcessing: false,
  clientSecretId: '',
};

const clientSecret = (state = clientSecretInitialState, action = null) => {
  if (action != null) {
    switch (action.type) {
      case fetchClientSecret.TRIGGER: return state;
      case fetchClientSecret.FULFILL: return state;
      case CLIENT_SECRET_DATA_RECEIVED: return { ...state, ...action.payload };
      case CLIENT_SECRET_PROCESSING: return { ...state, isClientSecretProcessing: action.payload };

      default:
    }
  }
  return state;
};

const reducer = combineReducers({
  basket,
  captureKey,
  clientSecret,
});

export default reducer;
