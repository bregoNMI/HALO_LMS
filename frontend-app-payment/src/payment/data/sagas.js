import {
  call, put, takeEvery, select, delay,
} from 'redux-saga/effects';
import { stopSubmit } from 'redux-form';
import { getReduxFormValidationErrors } from './utils';
import { MESSAGE_TYPES } from '../../feedback/data/constants';

// Actions
import {
  basketDataReceived,
  basketProcessing,
  captureKeyDataReceived,
  clientSecretDataReceived,
  captureKeyProcessing,
  CAPTURE_KEY_START_TIMEOUT,
  captureKeyStartTimeout,
  microformStatus,
  fetchBasket,
  addCoupon,
  removeCoupon,
  updateQuantity,
  submitPayment,
  fetchCaptureKey,
  clientSecretProcessing,
  fetchClientSecret,
} from './actions';

import { STATUS_LOADING } from '../checkout/payment-form/flex-microform/constants';

// Sagas
import { handleErrors, handleMessages, clearMessages } from '../../feedback';

// Services
import * as PaymentApiService from './service';
import { checkoutWithToken } from '../payment-methods/cybersource';
import { checkout as checkoutPaypal } from '../payment-methods/paypal';
import { checkout as checkoutApplePay } from '../payment-methods/apple-pay';
import { checkout as checkoutStripe } from '../payment-methods/stripe';

export const paymentMethods = {
  cybersource: checkoutWithToken,
  paypal: checkoutPaypal,
  'apple-pay': checkoutApplePay,
  stripe: checkoutStripe,
};

function* isBasketProcessing() {
  return yield select(state => state.payment.basket.isBasketProcessing);
}

function* isCaptureKeyProcessing() {
  return yield select(state => state.payment.captureKey.isCaptureKeyProcessing);
}

function* isClientSecretProcessing() {
  return yield select(state => state.payment.clientSecret.isClientSecretProcessing);
}

export function* handleReduxFormValidationErrors(error) {
  if (error.fieldErrors) {
    const fieldErrors = getReduxFormValidationErrors(error);
    yield put(stopSubmit('payment', fieldErrors));
  }
}

export function* handleDiscountCheck() {
  const basket = yield select(state => state.payment.basket);
  if (basket.products.length === 1) {
    const [product] = basket.products;
    const { courseKey } = product;
    if (product.productType === 'Seat') {
      const discount = yield call(
        PaymentApiService.getDiscountData,
        courseKey,
      );
      if (discount.discount_applicable) {
        const result = yield call(
          PaymentApiService.getBasket,
          discount.jwt,
        );
        result.discountJwt = discount.jwt;
        yield put(basketDataReceived(result));
        yield call(handleMessages, result.messages, false, window.location.search);
      }
    }
  }
}

export function* handleFetchBasket() {
  if (yield isBasketProcessing()) {
    // Do nothing if there is a request currently in flight to get or modify the basket
    return;
  }

  try {
    yield put(basketProcessing(true)); // we are going to modify the basket, don't make changes
    const result = yield call(PaymentApiService.getBasket);
    yield put(basketDataReceived(result)); // update redux store with basket data
    yield call(handleMessages, result.messages, true, window.location.search);
    yield call(handleDiscountCheck); // check if a discount should be added to the basket
  } catch (error) {
    yield call(handleErrors, error, true);
    if (error.basket) {
      yield put(basketDataReceived(error.basket)); // update redux store with basket data
      yield call(handleDiscountCheck); // check if a discount should be added to the basket
    }
  } finally {
    yield put(basketProcessing(false)); // we are done modifying the basket
    yield put(fetchBasket.fulfill()); // mark the basket as finished loading
  }
}

export function* handleCaptureKeyTimeout() {
  // Start at the 12min mark to leave 1 min of buffer on the 15min timeout
  yield delay(12 * 60 * 1000);
  yield call(
    handleMessages,
    [{
      code: 'capture-key-2mins-message',
      messageType: MESSAGE_TYPES.INFO,
    }],
    true, // Clear other messages
    window.location.search,
  );

  yield delay(1 * 60 * 1000);
  yield call(
    handleMessages,
    [{
      code: 'capture-key-1min-message',
      messageType: MESSAGE_TYPES.INFO,
    }],
    true, // Clear other messages
    window.location.search,
  );

  yield delay(1 * 60 * 1000);
  yield put(clearMessages());
  yield put(fetchCaptureKey());
}

/**
 * Redux Saga for getting the capture context for a cybersource payment
 */
export function* handleFetchCaptureKey() {
  if (yield isCaptureKeyProcessing()) {
    // Do nothing if there is a request currently in flight
    return;
  }

  try {
    yield put(captureKeyProcessing(true)); // we are waiting for a capture key
    yield put(microformStatus(STATUS_LOADING)); // we are refreshing the capture key
    const result = yield call(PaymentApiService.getCaptureKey);
    yield put(captureKeyDataReceived(result)); // update redux store with capture key data
    yield put(captureKeyStartTimeout());
    // only start the timer if we're using the capture key
  } catch (error) {
    yield call(handleErrors, error, true);
  } finally {
    yield put(captureKeyProcessing(false)); // we are done capture key
    yield put(fetchCaptureKey.fulfill()); // mark the capture key as finished loading
  }
}

/**
 * Redux saga for getting the client secret key for a Stripe payment
 */
export function* handleFetchClientSecret() {
  if (yield isClientSecretProcessing()) {
    return;
  }

  try {
    yield put(clientSecretProcessing(true));
    // TODO: possibly add status for stripe elements loading?
    const result = yield call(PaymentApiService.getClientSecret);
    yield put(clientSecretDataReceived(result));
  } catch (error) {
    yield call(handleErrors, error, true);
  } finally {
    yield put(clientSecretProcessing(false));
    yield put(fetchClientSecret.fulfill());
  }
}

/**
 * Many of the handlers here have identical error handling, and are also all processing the same
 * sort of response object (a basket).  This helper is just here to dedupe that code, since its
 * all identical.
 */
export function* performBasketOperation(operation, ...operationArgs) {
  if (yield isBasketProcessing()) {
    return;
  }

  try {
    yield put(basketProcessing(true));
    const result = yield call(operation, ...operationArgs);
    yield put(basketDataReceived(result));
    yield call(handleMessages, result.messages, true, window.location.search);
  } catch (error) {
    yield call(handleErrors, error, true);
    if (error.basket) {
      yield put(basketDataReceived(error.basket));
    }
  } finally {
    yield put(basketProcessing(false));
  }
}

export function* handleAddCoupon({ payload }) {
  yield call(performBasketOperation, PaymentApiService.postCoupon, payload.code);
}

export function* handleRemoveCoupon({ payload }) {
  yield call(performBasketOperation, PaymentApiService.deleteCoupon, payload.id);
}

export function* handleUpdateQuantity({ payload }) {
  yield call(performBasketOperation, PaymentApiService.postQuantity, payload);
}

export function* handleSubmitPayment({ payload }) {
  if (yield isBasketProcessing()) {
    return;
  }

  const { method, ...paymentArgs } = payload;
  try {
    yield put(basketProcessing(true));
    yield put(clearMessages()); // Don't leave messages floating on the page after clicking submit
    yield put(submitPayment.request());
    const paymentMethodCheckout = paymentMethods[method];
    const basket = yield select(state => ({ ...state.payment.basket }));
    yield call(paymentMethodCheckout, basket, paymentArgs);
    yield put(submitPayment.success());
  } catch (error) {
    // Do not handle errors on user aborted actions
    if (!error.aborted) {
      // Client side generated errors are simple error objects.  If we have one, wrap it in the
      // same format the API uses.
      if (error.code) {
        yield call(handleErrors, { messages: [error] }, true);
      } else {
        yield call(handleErrors, error, true);
        yield call(handleReduxFormValidationErrors, error);
      }

      if (error.basket) {
        yield put(basketDataReceived(error.basket));
        yield call(handleDiscountCheck);
      }
    }
  } finally {
    yield put(basketProcessing(false));
    yield put(submitPayment.fulfill());
  }
}

export default function* saga() {
  yield takeEvery(fetchCaptureKey.TRIGGER, handleFetchCaptureKey);
  yield takeEvery(CAPTURE_KEY_START_TIMEOUT, handleCaptureKeyTimeout);
  yield takeEvery(fetchClientSecret.TRIGGER, handleFetchClientSecret);
  yield takeEvery(fetchBasket.TRIGGER, handleFetchBasket);
  yield takeEvery(addCoupon.TRIGGER, handleAddCoupon);
  yield takeEvery(removeCoupon.TRIGGER, handleRemoveCoupon);
  yield takeEvery(updateQuantity.TRIGGER, handleUpdateQuantity);
  yield takeEvery(submitPayment.TRIGGER, handleSubmitPayment);
}
