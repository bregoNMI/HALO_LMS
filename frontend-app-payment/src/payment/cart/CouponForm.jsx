import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Button, Form } from '@openedx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';

import { addCoupon, removeCoupon } from '../data/actions';
import LocalizedPrice from './LocalizedPrice';

const renderMuted = txt => <span className="text-muted">{txt}</span>;

class CouponForm extends Component {
  constructor(props) {
    super(props);

    this.handleAddSubmit = this.handleAddSubmit.bind(this);
    this.handleRemoveSubmit = this.handleRemoveSubmit.bind(this);
  }

  handleAddSubmit(event) {
    event.preventDefault();
    this.props.addCoupon({ code: event.target.elements.couponField.value });
  }

  handleSubmitButtonClick() {
    sendTrackEvent(
      'edx.bi.ecommerce.basket.voucher_applied',
      {
        type: 'click',
        category: 'voucher-application',
      },
    );
  }

  handleRemoveSubmit(event) {
    event.preventDefault();
    this.props.removeCoupon({ id: this.props.id });
  }

  renderAdd() {
    const { code, isBasketProcessing } = this.props;
    const id = 'couponField';

    return (
      <form onSubmit={this.handleAddSubmit} className="summary-row d-flex align-items-end">
        <Form.Group controlId={id} invalid={false} className="mb-0 mr-2">
          <label className="h6 d-block" htmlFor={id}>
            <FormattedMessage
              id="payment.coupon.label"
              defaultMessage="Add coupon code (optional)"
              description="Label for the add coupon form"
            />
          </label>
          <Form.Control name={id} id={id} type="text" defaultValue={code} data-testid={id} />
        </Form.Group>
        <Button
          disabled={isBasketProcessing}
          variant="outline-primary"
          type="submit"
          onClick={this.handleSubmitButtonClick}
        >
          <FormattedMessage
            id="payment.coupon.submit"
            defaultMessage="Apply"
            description="Submit button for the add coupon form"
          />
        </Button>
      </form>
    );
  }

  renderCouponMessage() {
    const { code, benefitValue, benefitType } = this.props;

    if (benefitType === 'Absolute') {
      return (
        <span className="text-muted">
          <FormattedMessage
            id="payment.coupon.benefit.absolute"
            defaultMessage="Coupon {code} applied for {amount} off"
            description="A coupon has been applied for a fixed currency discount, like $10.  Currency symbol will already be provided."
            values={{
              code,
              amount: <LocalizedPrice amount={benefitValue} />,
            }}
          />
        </span>
      );
    }

    if (benefitType === 'Percentage') {
      return (
        <FormattedMessage
          id="payment.coupon.benefit.percentage"
          defaultMessage="Coupon {code} applied for {amount}% off"
          description="A coupon has been applied for a percentage discount, like 10%.  Please place the % symbol as appropriate."
          values={{
            code,
            amount: benefitValue,
          }}
        >
          {renderMuted}
        </FormattedMessage>
      );
    }

    return (
      <FormattedMessage
        id="payment.coupon.benefit.default"
        defaultMessage="Coupon {code} applied"
        description="A coupon has been applied."
        values={{ code }}
      >
        {renderMuted}
      </FormattedMessage>
    );
  }

  renderRemove() {
    return (
      <form onSubmit={this.handleRemoveSubmit} className="d-flex align-items-center mb-3">
        {this.renderCouponMessage()}
        <Button
          className="display-inline p-0 pl-3 border-0"
          variant="link"
          type="submit"
          disabled={this.props.isBasketProcessing}
        >
          <FormattedMessage
            id="payment.coupon.remove"
            defaultMessage="Remove"
            description="Submit button to remove a coupon"
          />
        </Button>
      </form>
    );
  }

  render() {
    if (this.props.id !== null) {
      return this.renderRemove();
    }

    return this.renderAdd();
  }
}

CouponForm.propTypes = {
  isBasketProcessing: PropTypes.bool,
  code: PropTypes.string,
  id: PropTypes.number,
  addCoupon: PropTypes.func.isRequired,
  removeCoupon: PropTypes.func.isRequired,
  benefitValue: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  benefitType: PropTypes.string,
};

CouponForm.defaultProps = {
  isBasketProcessing: false,
  code: '',
  id: null,
  benefitValue: null,
  benefitType: null,
};

const mapStateToProps = (state) => {
  const { basket } = state.payment;
  const coupon = basket.coupons && basket.coupons.length ? basket.coupons[0] : {};

  return {
    ...coupon,
    isBasketProcessing: basket.isBasketProcessing,
  };
};

export default connect(
  mapStateToProps,
  {
    addCoupon,
    removeCoupon,
  },
)(CouponForm);
