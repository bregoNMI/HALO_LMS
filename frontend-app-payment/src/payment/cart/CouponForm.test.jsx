import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import defaultsDeep from 'lodash.defaultsdeep';
import configureMockStore from 'redux-mock-store';
import { IntlProvider } from '@edx/frontend-platform/i18n';
import { sendTrackEvent } from '@edx/frontend-platform/analytics';

import CouponForm from './CouponForm';
import { addCoupon, removeCoupon } from '../data/actions';

jest.mock('@edx/frontend-platform/analytics', () => ({
  sendTrackEvent: jest.fn(),
}));

const mockStore = configureMockStore();

const baseState = {
  payment: {
    basket: {
      coupons: [],
      isBasketProcessing: false,
    },
  },
};

const baseCoupon = {
  id: 1234,
  code: 'COUPONCODE',
  benefitValue: '10',
  benefitType: 'Absolute',
};

const renderCouponFormWithState = store => (
  <IntlProvider locale="en">
    <Provider store={store}>
      <CouponForm />
    </Provider>
  </IntlProvider>
);

describe('CouponForm', () => {
  it('should render an absolute value coupon', () => {
    const component = renderCouponFormWithState(mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [{ ...baseCoupon, benefitType: 'Absolute' }],
        },
      },
    }, baseState)));
    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });

  it('should render a percentage value coupon', () => {
    const component = renderCouponFormWithState(mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [{ ...baseCoupon, benefitType: 'Percentage' }],
        },
      },
    }, baseState)));
    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });

  it('should render a coupon even if no known benefitType is supplied', () => {
    const component = renderCouponFormWithState(mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [{ ...baseCoupon, benefitType: 'Barter' }],
        },
      },
    }, baseState)));
    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });

  it('should render a form when there is no coupon', () => {
    const component = renderCouponFormWithState(mockStore(baseState));
    const { container: tree } = render(component);
    expect(tree).toMatchSnapshot();
  });

  describe('Add Coupon', () => {
    it('sends a track event on clicking the add submit coupon button', () => {
      const store = mockStore(baseState);
      const component = renderCouponFormWithState(store);
      const { container } = render(component);
      fireEvent.click(container.querySelector('button[type="submit"]'));

      expect(sendTrackEvent).toHaveBeenCalledWith(
        'edx.bi.ecommerce.basket.voucher_applied',
        { type: 'click', category: 'voucher-application' },
      );
    });

    it('fires an add coupon action on submit', () => {
      const store = mockStore(baseState);
      const component = renderCouponFormWithState(store);
      const { container } = render(component);
      fireEvent.submit(container.querySelector('button[type="submit"]'), {
        target: { elements: { couponField: { value: 'testvalue' } } },
      });

      expect(store.getActions()).toEqual([
        addCoupon({ code: 'testvalue' }),
      ]);
    });
  });

  describe('Remove Coupon', () => {
    const store = mockStore(defaultsDeep({
      payment: {
        basket: {
          coupons: [baseCoupon],
        },
      },
    }, baseState));
    const component = renderCouponFormWithState(store);

    it('fires an remove coupon action on submit', () => {
      const { container } = render(component);
      fireEvent.submit(container.querySelector('form'));

      expect(store.getActions()).toEqual([
        removeCoupon({ id: 1234 }),
      ]);
    });
  });
});
