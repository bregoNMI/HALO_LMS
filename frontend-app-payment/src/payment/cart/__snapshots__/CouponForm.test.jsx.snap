// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CouponForm should render a coupon even if no known benefitType is supplied 1`] = `
<div>
  <form
    class="d-flex align-items-center mb-3"
  >
    <span
      class="text-muted"
    >
      Coupon COUPONCODE applied
    </span>
    <button
      class="display-inline p-0 pl-3 border-0 btn btn-link"
      type="submit"
    >
      Remove
    </button>
  </form>
</div>
`;

exports[`CouponForm should render a form when there is no coupon 1`] = `
<div>
  <form
    class="summary-row d-flex align-items-end"
  >
    <div
      class="pgn__form-group mb-0 mr-2"
    >
      <label
        class="h6 d-block"
        for="couponField"
      >
        Add coupon code (optional)
      </label>
      <div
        class="pgn__form-control-decorator-group"
      >
        <input
          class="form-control"
          data-testid="couponField"
          id="couponField"
          name="couponField"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="btn btn-outline-primary"
      type="submit"
    >
      Apply
    </button>
  </form>
</div>
`;

exports[`CouponForm should render a percentage value coupon 1`] = `
<div>
  <form
    class="d-flex align-items-center mb-3"
  >
    <span
      class="text-muted"
    >
      Coupon COUPONCODE applied for 10% off
    </span>
    <button
      class="display-inline p-0 pl-3 border-0 btn btn-link"
      type="submit"
    >
      Remove
    </button>
  </form>
</div>
`;

exports[`CouponForm should render an absolute value coupon 1`] = `
<div>
  <form
    class="d-flex align-items-center mb-3"
  >
    <span
      class="text-muted"
    >
      Coupon COUPONCODE applied for 
      $10.00
       off
    </span>
    <button
      class="display-inline p-0 pl-3 border-0 btn btn-link"
      type="submit"
    >
      Remove
    </button>
  </form>
</div>
`;
